src/python/foglamp/device/ingest.py

```diff
@@ -0,0 +1,497 @@
+import asyncpg
+import dateutil.parser
+import json
+
+from foglamp import logger
+from foglamp import statistics
+
+
+__author__ = "Terris Linenbach"
+__copyright__ = "Copyright (c) 2017 OSIsoft, LLC"
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
+
+_LOGGER = logger.setup(__name__)  # type: logging.Logger
+# _LOGGER = logger.setup(__name__, level=logging.DEBUG)  # type: logging.Logger
+# _LOGGER = logger.setup(__name__, destination=logger.CONSOLE, level=logging.DEBUG)
+
+_STATISTICS_WRITE_FREQUENCY_SECONDS = 5
+
+
+class Ingest(object):
+    """Adds sensor readings to FogLAMP
+
+    Also tracks readings-related statistics.
+
+    Readings are added to a configurable number of queues. These queues are processed
+    concurrently. Each queue is assigned to a database connection. Queued items are
+    batched into a single insert transaction. The size of these batches have a
+    configurable maximum and minimum.
+    """
+
+    # Class attributes
+    _readings_stats = 0  # type: int
+    """Number of readings accepted before statistics were flushed to storage"""
+
+    _discarded_readings_stats = 0  # type: int
+    """Number of readings rejected before statistics were flushed to storage"""
+
+    _write_statistics_task = None  # type: asyncio.Task
+    """asyncio task for :meth:`_write_statistics`"""
+
+    _write_statistics_sleep_task = None  # type: asyncio.Task
+    """asyncio task for asyncio.sleep"""
+
+    _stop = False
+    """Set to true when the server needs to stop"""
+
+    _started = False
+    """True when the server has been started"""
+
+    _readings_queues = None  # type: List[asyncio.Queue]
+    """insert objects are added to these queues"""
+
+    _current_readings_queue_index = 0
+    """Which queue to insert into next"""
+
+    _insert_readings_tasks = None  # type: List[asyncio.Task]
+    """asyncio tasks for :meth:`_insert_readings`"""
+
+    _queue_events = None  # type: List[asyncio.Event]
+
+    _insert_readings_wait_tasks = None  # type: List[asyncio.Task]
+    """asyncio tasks for asyncio.Queue.get called by :meth:`_insert_readings`"""
+
+    # Configuration
+    _num_readings_queues = 1
+    """Maximum number of insert queues. Each queue has its own database connection."""
+
+    _max_idle_db_connection_seconds = 180
+    """Close database connections when idle for this number of seconds"""
+
+    _min_readings_batch_size = 50
+    """Preferred minimum number of rows in a batch of inserts"""
+
+    _max_readings_batch_size = 150
+    """Maximum number of rows in a batch of inserts"""
+    
+    _max_readings_queue_size = 4*_max_readings_batch_size
+    """Maximum number of items in a queue"""
+
+    _readings_batch_yield_items = 50
+    """While creating a batch, yield to other tasks after this taking this many
+    items from the queue"""
+
+    _readings_batch_wait_seconds = 1
+    """Number of seconds to wait for a queue to reach the minimum batch size"""
+
+    _max_insert_readings_batch_attempts = 60
```

Comment:
Though it shall be configurable; but I think its too high ... 60 attempts!

Reply:
In the case of a connection failure, there is a 1 second sleep in between attempts. So 60 means to try to connect for 1 minute. I can change it to 30.